traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go

```diff
@@ -0,0 +1,414 @@
+// is connected when the federations/:id/deliveryservice links a federation
+// Note: cdns and deliveryservies have a 1-1 relationship
+func (fed *TOCDNFederation) Create() (error, tc.ApiErrorType) {
+
+	// Deliveryservice IDs should not be included on create.
+	if fed.DeliveryServiceIDs != nil {
+		fed.DsId = nil
+		fed.XmlId = nil
+		fed.DeliveryServiceIDs = nil
+	}
+
+	// Boilerplate code below
+	resultRows, err := fed.ReqInfo.Tx.NamedQuery(insertQuery(), fed)
+	if err != nil {
+		return parseQueryError(err, "create")
+	}
+	defer resultRows.Close()
+
+	var id int
+	var lastUpdated tc.TimeNoMod
+
+	rowsAffected := 0
+	for resultRows.Next() {
+		rowsAffected++
+		if err = resultRows.Scan(&id, &lastUpdated); err != nil {
+			log.Error.Printf("could not scan id and last_updated from insert: %s\n", err)
+			return tc.DBError, tc.SystemError
+		}
+	}
+	if rowsAffected == 0 {
+		err = errors.New("no federation was inserted, no id was returned")
+		log.Errorln(err)
+		return tc.DBError, tc.SystemError
+	} else if rowsAffected > 1 {
+		err = errors.New("too many ids returned from fed insert")
+		log.Errorln(err)
+		return tc.DBError, tc.SystemError
+	}
+	fed.SetKeys(map[string]interface{}{"id": id})
+	fed.LastUpdated = &lastUpdated
+
+	return nil, tc.NoError
+}
+
+// returning true indicates the data related to the given tenantID should be visible
+// `tenantIDs` is presumed to be unsorted, and a nil tenantID is viewable by everyone
+func checkTenancy(tenantID *int, tenantIDs []int) bool {
+	if tenantID == nil {
+		return true
+	}
+	for _, id := range tenantIDs {
+		if id == *tenantID {
+			return true
+		}
+	}
+	return false
+}
+
+func (fed *TOCDNFederation) Read(parameters map[string]string) ([]interface{}, []error, tc.ApiErrorType) {
+
+	// Cannot perform query on tenantID while "rows" aren't closed (limitation of
+	// psql), so we need to get the valid tenentIDs ahead of time.
+	tenantIDs, err := tenant.GetUserTenantIDListTx(fed.ReqInfo.Tx.Tx, fed.ReqInfo.User.TenantID)
+	if err != nil {
+		log.Errorf("getting tenant list for user: %v\n", err)
+		return nil, []error{tc.DBError}, tc.SystemError
+	}
+
+	var query string
+	_, id := parameters["id"]
+	queryParamsToQueryCols := map[string]dbhelpers.WhereColumnInfo{
+		"id": dbhelpers.WhereColumnInfo{Column: "federation.id", Checker: api.IsInt},
+	}
+	if !id {
+		queryParamsToQueryCols["name"] = dbhelpers.WhereColumnInfo{Column: "cdn.name", Checker: nil}
+	}
+	where, orderBy, queryValues, errs := dbhelpers.BuildWhereAndOrderBy(parameters, queryParamsToQueryCols)
+
+	if id {
+		// Can't use `AddTenancyCheck` for id because then we won't know what caused
+		// an empty response, so the tenancy check will be performed below.
+		query = selectByID()
+	} else { // searching by name
+		query = selectByCDNName()
+		where, queryValues = dbhelpers.AddTenancyCheck(where, queryValues, "(tenant_id", tenantIDs)
+		where += " OR tenant_id IS NULL)"
```

Comment:
I think this null tenant_id check can be removed because we now enforce non-null tenancy everywhere, but I won't hold this up because of that.